op-node/sources/fallback_client.go

```diff
@@ -0,0 +1,280 @@
+package sources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/metrics"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// FallbackClient is an RPC client, it can automatically switch to the next l1 endpoint
+// when there is a problem with the current l1 endpoint
+// and automatically switch back after the first l1 endpoint recovers.
+type FallbackClient struct {
```

Comment:
This may be conflicting with https://github.com/ethereum-optimism/optimism/pull/7029, but the fallback client should be layered onto the client & be easy to create.

Reply:
It seems that #7029 was not merged, but rather closed. Right now, I can't be sure what it will turn out to be like.

> the fallback client should be layered onto the client & be easy to create.

I tried to achieve this goal, but since the fallback client needs to create multiple underlying RPC clients instead of just adding new functions to a single client, the situation is a bit more complicated. Currently I try to implement the client.RPC interface and wrap the original client.NewRPC when multiple addresses are configured. See [fallbackClientWrap ](https://github.com/ethereum-optimism/optimism/pull/7311/files#diff-a27156a67289c4da44b8746353b897cde1e4fb77737af85838ad9ca91923ef12R197) method.
This is the easiest way I can think of to create it. If you have any better ideas, please let me know.